### PR TITLE
refactor(connlib): move side-effects up the callstack

### PR DESCRIPTION
**Pull request by thomaseizinger**
This refactoring series removes the `channel` property from `Peer`. The idea here is that we want to move side-effects _up_ the callstack. Side-effects, esp. network access usually requires ownership over a certain resource like the data channel or the TUN device. It is easier to handle this ownership if we never give out references to these resources but instead keep them in the upper layer and keep the rest of the code side-effect free.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
@conectado This is now passing CI and ready for review :)

Let me know if you are happy to merge in this state or whether I should include more commits. The things I am planning to do next are:

- Attempt to remove `Arc` from `Peer` and use `&mut self`.
- Maybe wrap `Arc<Datachannel>` in a new-type and co-locate its read-buffer with it.

File: rust/connlib/tunnel/src/peer.rs [hunk 1]
```diff
@@ -142,21 +134,13 @@ where
         let mut buf = [0u8; MAX_SCRATCH_SPACE];
 
         let packet = match self.tunnel.lock().update_timers(&mut buf) {
-            TunnResult::Done => return Ok(()),
+            TunnResult::Done => return Ok(Bytes::new()),
```
Comment (conectado) [inline rust/connlib/tunnel/src/peer.rs]: I'd rather return a `Result<Option<Bytes>>` so that we don't send empty packets.

We can also check if the packet is empty and not send it.

Comment (conectado) [reply rust/connlib/tunnel/src/peer.rs]: This is because afaik in `webrtc` empty packets are a thing, so something could be sent to the reciever.

File: rust/connlib/tunnel/src/peer.rs [hunk 2]
```diff
@@ -220,18 +204,44 @@ where
 
             packet.update_checksum();
         }
-        let packet = match self.tunnel.lock().encapsulate(packet.packet_mut(), buf) {
-            TunnResult::Done => return Ok(()),
+        let packet = match self.tunnel.lock().encapsulate(packet.packet(), buf) {
+            TunnResult::Done => return Ok(Bytes::new()),
```
Comment (conectado) [inline rust/connlib/tunnel/src/peer.rs]: Same here.

File: rust/connlib/tunnel/src/lib.rs [hunk 3]
```diff
@@ -141,7 +142,8 @@ pub struct Tunnel<CB: Callbacks, TRoleState: RoleState> {
     rate_limiter: Arc<RateLimiter>,
     private_key: StaticSecret,
     public_key: PublicKey,
-    peers_by_ip: RwLock<IpNetworkTable<Arc<Peer<TRoleState::Id>>>>,
+    #[allow(clippy::type_complexity)]
+    peers_by_ip: RwLock<IpNetworkTable<(Arc<Peer<TRoleState::Id>>, Arc<DataChannel>)>>,
```
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: Introduce a new type, I think we tidy this up here and we remove the clippy exception

```rs
ConnectedPeer<T> {
  peer: Arc<Peer<T>>,
  channel: Arc<DataChannel>
}
```

Even if we are going to get rid of it latter I'd rather merge it like this.

File: rust/connlib/tunnel/src/control_protocol/client.rs [hunk 4]
```diff
@@ -129,13 +133,19 @@ where
...
                 };
 
-                d.on_close(on_dc_close_handler(index, gateway_id, tunnel.stop_peer_command_sender.clone()));
+                d.on_close(on_dc_close_handler(
+                    index,
+                    gateway_id,
+                    tunnel.stop_peer_command_sender.clone(),
+                ));
+                let d = d.detach().await.expect(
+                    "only fails if not opened or not enabled, both of which are always true for us",
```
Comment (conectado) [inline rust/connlib/tunnel/src/control_protocol/client.rs:142]: Sorry, this is not part of this PR but here we will panic in a race condition where the channel closes while we are executing this callback. It's unlikely but I'm not sure it's impossible.

**Review by conectado — COMMENTED** [review]
Looks good to me, left a few comments.

I'd rather merge it, including, if needed, the commented changes, before including more commits.

File: rust/connlib/tunnel/src/control_protocol/client.rs (hunk 4 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/control_protocol/client.rs:142]: Do we care enough to handle it "properly" for now?

Comment (conectado) [reply rust/connlib/tunnel/src/control_protocol/client.rs:142]: I do think we should at least not panic and remove the peer.

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/control_protocol/client.rs:142]: Where do you see this race condition? The source only mentions an error if it is called before opening: https://docs.rs/webrtc/latest/src/webrtc/data_channel/mod.rs.html#400-413

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/control_protocol/client.rs:142]: I cannot find any place in the source code where this is set back to `None` after it has been opened.

Comment (conectado) [reply rust/connlib/tunnel/src/control_protocol/client.rs:142]: My bad, I thought that it might have set to `None` after closing.

**Review by conectado — APPROVED** [review]
LGTM
